src/storage/compression/bitpacking.cpp

```diff
@@ -0,0 +1,631 @@
+	template <class T>
+	inline static bitpacking_width_t MinimumBitWidth(T *values, idx_t count) {
+		return FindMinimumBitWidth<T, BYTE_ALIGNED>(values, count);
+	}
+
+private:
+	template <class T>
+	static bitpacking_width_t MinimumBitWidth(T min_value, T max_value) {
+		bitpacking_width_t required_bits;
+
+		if (std::is_signed<T>::value) {
+			if (min_value == NumericLimits<T>::Minimum()) {
+				// handle special case of the minimal value, as it cannot be negated like all other values.
+				return sizeof(T) * 8;
+			} else {
+				max_value = MaxValue((T)-min_value, max_value);
+			}
+		}
+
+		if (max_value == 0) {
+			return 0;
+		}
+
+		if (std::is_signed<T>::value) {
+			required_bits = 1;
+		} else {
+			required_bits = 0;
+		}
+
+		while (max_value) {
+			required_bits++;
+			max_value >>= 1;
+		}
+
+		return GetEffectiveWidth<T>(required_bits);
+	}
+
+	template <class T, bool round_to_next_byte = false>
+	static bitpacking_width_t FindMinimumBitWidth(T *values, idx_t count) {
+		T min_value = values[0];
+		T max_value = values[0];
+
+		for (idx_t i = 1; i < count; i++) {
+			if (values[i] > max_value) {
+				max_value = values[i];
+			}
+
+			if (std::is_signed<T>::value) {
+				if (values[i] < min_value) {
+					min_value = values[i];
+				}
+			}
+		}
+
+		bitpacking_width_t calc_width = MinimumBitWidth<T>(std::is_signed<T>::value ? min_value : 0, max_value);
+
+		// Assert results are correct
+#ifdef DEBUG
+		if (calc_width < sizeof(T) * 8 && calc_width != 0) {
+			if (std::is_signed<T>::value) {
+				D_ASSERT((int64_t)max_value <= (int64_t)(1L << (calc_width - 1)) - 1);
+				D_ASSERT((int64_t)min_value >= (int64_t)(-1 * ((1L << (calc_width - 1)) - 1) - 1));
+			} else {
+				D_ASSERT((uint64_t)max_value <= (uint64_t)(1L << (calc_width)) - 1);
+			}
+		}
+#endif
+		if (round_to_next_byte) {
+			return (calc_width / 8 + (calc_width % 8 != 0)) * 8;
+		} else {
+			return calc_width;
+		}
+	}
+
+	template <class T>
+	static void UnPackGroup(data_ptr_t dst, data_ptr_t src, bitpacking_width_t width) {
+		if (std::is_same<T, uint8_t>::value || std::is_same<T, int8_t>::value) {
+			duckdb_fastpforlib::fastunpack((const uint8_t *)src, (uint8_t *)dst, (uint32_t)width);
+		} else if (std::is_same<T, uint16_t>::value || std::is_same<T, int16_t>::value) {
+			duckdb_fastpforlib::fastunpack((const uint16_t *)src, (uint16_t *)dst, (uint32_t)width);
+		} else if (std::is_same<T, uint32_t>::value || std::is_same<T, int32_t>::value) {
+			duckdb_fastpforlib::fastunpack((const uint32_t *)src, (uint32_t *)dst, (uint32_t)width);
+		} else if (std::is_same<T, uint64_t>::value || std::is_same<T, int64_t>::value) {
+			duckdb_fastpforlib::fastunpack((const uint32_t *)src, (uint64_t *)dst, (uint32_t)width);
+		} else {
+			throw InternalException("Unsupported type found in bitpacking.");
+		}
+
+		if (NumericLimits<T>::IsSigned() && width > 0 && width < sizeof(T) * 8) {
+			SignExtend<T>(dst, width);
```

Comment:
The SignExtend is only required for columns that hold negative numbers, right? Perhaps we could check the stats to see if this is required?